### ISSUE TITLE
BC37255: Combined length of user strings used by the program exceeds allowed limit. Try to decrease use of string or XML literals

### ISSUE THREAD

**Issue by vsfeedback** [dotnet/roslyn#25136]
I receive this error on my datasets since I've installed the latest VS2017. I have big tables, and think the MSDataSetGenerator generates code different, so now the compiler can't handle the big generated strings anymore?

There is not much information about the error message BC37255.

Can you help?

_This issue has been moved from https://developercommunity.visualstudio.com/content/problem/192728/bc37255-combined-length-of-user-strings-used-by-th.html
VSTS ticketId: 561492_
_These are the original issue comments:_
(no comments)
_These are the original issue solutions:_
(no solutions)

**Comment by jaredpar** [comment]
Can you provide a bit more information here? There are limits on how big strings and XML literals can be in an application. But it's hard to see from the issue why this is happening here and what can be done to alleviate it.

**Comment by gr8tushar** [comment]
Hi,
I am also facing the same issue. It was absolutely working fine in 2015. But in VS 2017, it is giving the above mentioned error.
We have about 6 projects in a solutions. One Web App and others as class libraries. Each project has a constant file. That file would have 100s and 1000s of static strings - be it queries, constants, cache keys, session keys, messages, etc. 

Do let me know in case you need more info on the same.

**Comment by rynowak** [comment]
hi @soiwill - sorry you're encountering this problem. We've implemented featured inside the Razor compiler to avoid this problem in the past, and it looks like either this is a new case, or we broke something in the 2.1 SDK.

Could I ask you to report this over at https://github.com/aspnet/AspNetCore/issues so we can follow up as this problem relates to ASP.NET Core?

I would be great to know the version of the runtime you're targeting as well as the exact SDK version.

**Comment by gafter** [comment]
Closing as fixed in 16.4 due to https://github.com/dotnet/roslyn/pull/37915.  Please reopen if that is not in fact fixed.

**Comment by tmat** [comment]
@gafter I think this might be a different issue than the one addressed by #37915, since BC37255 is report by the metadata writer when the assembly user string heap overflows. 

In VS 2015 the compiler did not correctly check the size of the strings and might have produced an invalid assembly if the total string size exceeded the limits allows by ECMA335 specification.

To avoid this error, I'd recommend to store large strings (or large quantities of strings) to embedded resources instead of string literals.

**Comment by BeelineDevelopers** [comment]
> Closing as fixed in 16.4 due to #37915. Please reopen if that is not in fact fixed.

16.4 Preview 4 still has this issue. It increases the maximum combined length of string over 16.3.8 but it doesn't solve the problem entirely. As you increase the number of strings in 16.4 preview 4 the issue comes back.

Is there a way to fix it altogether?

Thanks in advance.

**Comment by gafter** [comment]
> Is there a way to fix it altogether?

This is a limitation of the specification for the assembly file format.  We would need to revise that to lift the limitation.

If you have large string literals that you need in your program, you can move them to resources so that they do not take up space in the string table.

**Comment by BeelineDevelopers** [comment]
> > Is there a way to fix it altogether?
> 
> This is a limitation of the specification for the assembly file format. We would need to revise that to lift the limitation.
> 
> If you have large string literals that you need in your program, you can move them to resources so that they do not take up space in the string table.

We do not have large strings but many strings across many .cshtml files and combining them all exceeds the threshold apparently.

Aside from defining it in the resources, which I assume will impact performance, are there other ways that won't affect performance but it still is able to compile and function?

What's involved in revising the limitation? Can it be configurable? What is it set at now?

Thanks in advance for your help.

**Comment by gafter** [comment]
The `#US` "user string" heap in an assembly has a "Size" that is stored as 4 bytes.  There are no plans for an overhaul of the assembly file format.  If we were to embark on a project of that scale, I would expect rollout of all of the affected tools to take several years.

The only other thing I can suggest is to break the program into several assemblies.

**Comment by BeelineDevelopers** [comment]
> The `#US` "user string" heap in an assembly has a "Size" that is stored as 4 bytes. There are no plans for an overhaul of the assembly file format. If we were to embark on a project of that scale, I would expect rollout of all of the affected tools to take several years.
> 
> The only other thing I can suggest is to break the program into several assemblies.

Thanks for your response.

If I had a .NET Core website with thousands of pages and each page would have variables defining the content of the site, for instance a title variable which is used in the <title> tag, <keywords> tag, <og> tags and in the <h1>, how would I set up the project(s) to get around the size limitation of the concatenation?

**Comment by CyrusNajmabadi** [comment]
Why not store them in external files (like resource variables) and have your code load it from that?  Then the strings aren't in your dlls themselves, but they're just effectively in xml files that can be effectively any size.

**Comment by CyrusNajmabadi** [comment]
Note: these questions are better suited to something like gitter/discord.

**Comment by BeelineDevelopers** [comment]
> Why not store them in external files (like standard .net resource files) and have your code load it from that? Then the strings aren't in your dlls themselves, but they're just effectively in xml files that can be effectively any size.

Thanks for the tip!

Is there any performance implications going that route (resource files) vs. having them in variables? In our scenario, even milliseconds count. I would think reading the file would be a hit to performance but maybe after that it's OK since it would be cached in memory?

**Comment by CyrusNajmabadi** [comment]
> Is there any performance implications

Yes.  Measure.

> In our scenario, even milliseconds count. 

Measure.

>  but maybe after that it's OK since it would be cached in memory?

Maybe.  Measure.   This is your domain, so you will need to evaluate what the performance is and what is acceptable.

**Comment by CyrusNajmabadi** [comment]
As mentioned above:  these questions are better suited to something like gitter/discord.  Try gitter.im/dotnet/roslyn  Or discord.gg/csharp

**Comment by BeelineDevelopers** [comment]
> > Is there any performance implications
> 
> Yes. Measure.
> 
> > In our scenario, even milliseconds count.
> 
> Measure.
> 
> > but maybe after that it's OK since it would be cached in memory?
> 
> Maybe. Measure. This is your domain, so you will need to evaluate what the performance is and what is acceptable.

True. 

But, in general, in your expert opinion, is there a performance degradation using resources vs. variables on a website which in your opinion is not negligible.

**Comment by CyrusNajmabadi** [comment]
As mentioned above: these questions are better suited to something like gitter/discord. Try gitter.im/dotnet/roslyn Or discord.gg/csharp

**Comment by dotcom9** [comment]
Still getting this error trying to build an ASP.NET Core 3.1 application in VS2019 16.4.2.

**Error CS8103 Combined length of user strings used by the program exceeds allowed limit. Try to decrease use of string literals.**

Microsoft Visual Studio Professional 2019
Version 16.4.2
VisualStudio.16.Release/16.4.2+29613.14
Microsoft .NET Framework
Version 4.8.03752

Installed Version: Professional

ASP.NET and Web Tools 2019 16.4.457.38025
ASP.NET and Web Tools 2019

ASP.NET Web Frameworks and Tools 2019 16.4.457.38025

**Comment by ferbalu22** [comment]
.net core >= 2.1 && < 3
This can be accomplished using the property RazorCompileOnBuild in the .csproj file:

<PropertyGroup>
  <TargetFramework>netcoreapp2.1</TargetFramework>
  <RazorCompileOnBuild>false</RazorCompileOnBuild>
  <RazorCompileOnPublish>false</RazorCompileOnPublish>
</PropertyGroup>

**Comment by gafter** [comment]
@ferbalu22 I think you are confirming that this is fixed in .net core 3.

**Comment by gr8tushar** [comment]
This issue is still not fixed. 
I have tried Visual Studio Pro 2019 (16.5.5) and .NET Core 3.1

As soon as I add few Views to an ASP.NET Core project, the build fails. 

Its strange that no enterprise has reported this. I fail to understand how can an enterprise product be migrated to .NET Core if this issue is not resolved. 
Having few hundred views in a enterprise project is a very common scenario.

I have started feeling that .NET Core is more of a playing ground for Microsoft. It's not meant or mature enough for enterprise development.
